osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsGasCommandExecutor.java

```diff
@@ -333,24 +494,360 @@ private Date readCaptureTime(
+        .toList();
+  }
+
+  private int getChannelWithoutException(final CosemObject object) {
+    try {
+      return object.getChannel();
+    } catch (final ObjectConfigException e) {
+      return -1;
+    }
+  }
+
+  private Map<CaptureObject, String> getScalerUnitsFromMeter(
+      final List<CaptureObject> captureObjects,
+      final DlmsConnectionManager conn,
+      final DlmsDevice device)
+      throws ProtocolAdapterException {
+    if (captureObjects.isEmpty()) {
+      return Map.of();
+    }
+
+    final Map<CaptureObject, String> captureObjectsWithScalerUnit = new HashMap<>();
+
+    final AttributeAddress[] scalerUnitAddresses = this.getScalerUnitAddresses(captureObjects);
+
+    conn.getDlmsMessageListener()
+        .setDescription(
+            String.format(
+                "GetPeriodicMeterReadsGas scaler units, retrieve attribute: %s",
+                JdlmsObjectToStringUtil.describeAttributes(scalerUnitAddresses)));
+
+    final List<GetResult> getResults =
+        this.dlmsHelper.getWithList(conn, device, scalerUnitAddresses);
+
+    if (getResults.stream().anyMatch(result -> result.getResultCode() != AccessResultCode.SUCCESS)
+        || getResults.size() != captureObjects.size()) {
+      throw new ProtocolAdapterException(
+          "Could not get all scaler units from meter: " + getResults);
+    }
+
+    final List<String> scalerUnits = this.readScalerUnits(getResults);
+
+    for (int i = 0; i < scalerUnits.size(); i++) {
+      captureObjectsWithScalerUnit.put(captureObjects.get(i), scalerUnits.get(i));
+    }
+
+    return captureObjectsWithScalerUnit;
+  }
+
+  private AttributeAddress[] getScalerUnitAddresses(final List<CaptureObject> captureObjects) {
+    return captureObjects.stream().map(this::getScalerUnitAddress).toArray(AttributeAddress[]::new);
+  }
+
+  private AttributeAddress getScalerUnitAddress(final CaptureObject captureObject) {
+    final CosemObject cosemObject = captureObject.getCosemObject();
+    return new AttributeAddress(
+        cosemObject.getClassId(),
+        new ObisCode(cosemObject.getObis()),
+        ExtendedRegisterAttribute.SCALER_UNIT.attributeId());
+  }
+
+  private List<String> readScalerUnits(final List<GetResult> getResultList)
+      throws ProtocolAdapterException {
+
+    final List<String> scalerUnits = new ArrayList<>();
+
+    for (final GetResult getResult : getResultList) {
+      final DataObject scalerUnitObject = getResult.getResultData();
+
+      scalerUnits.add(
+          this.dlmsHelper.getScalerUnit(
+              scalerUnitObject, "get scaler unit for periodic meter reads"));
+    }
+
+    return scalerUnits;
+  }
+
+  /**
+   * get device model code from a comma seperated list per channel index 1-4 is channel 1-4 and
+   * index 0 is device model code of master device
+   */
+  public String getDeviceModelCodeOfChannel(final String codes, final int channel) {

```

Comment:
This function is no longer used? I guess it is replaced by CombinedDeviceModelCode.parse?

Reply:
Yes, thats correct. I removed this method.